### ISSUE TITLE
Menu highlighting fails when a menu is reopened

### ISSUE THREAD

**Issue by spl237** [WayfireWM/wayfire#1814]
**Describe the bug**
If a menu is opened and the mouse is used to activate an item in that menu, the next time that menu is opened, the item which was activated does not get a highlight when the mouse is over it.

**To Reproduce**
1. Run a GTK application with menus
2. Open a menu and click an item in the menu with the mouse
3. Reopen the same menu and hover the mouse over the same item you previously clicked
4. All items in the menu will highlight when the mouse is over them except the one previously clicked

**Expected behavior**
All items in the menu should highlight when the mouse is over them

**Wayfire version**
0.7.4-3, Debian bookworm

**Comment by ammen99** [comment]
There have been many changes since Wayfire 0.7.4 on the master branch (including related to popup menus), you might want to check whether the bug has been fixed there.

**Comment by ammen99** [comment]
I can reproduce the issue with gnome-terminal. But the same issue is present in Sway, I think this might simply be a gtk bug.

**Comment by spl237** [comment]
I've done some more digging into this today; I don't think it is a GTK bug.

From what I can see, in the circumstance where a menu item has been clicked and the menu is reopened, GTK never even receives an enter-notify event when the mouse is moved back into the menu item; the event is never generated. Tracing through the code, the problem appears to be that the seat pointer should be generating the enter notify event, but never does so in this particular circumstance.

As far as I can see, most of the GTK code is common to X and Wayland (and this works fine in X), and the GDK Wayland backend is merely passing on this event from the seat; I can't see anywhere it is blocking or not generating it.

**Comment by ammen99** [comment]
Does the enter event show up with `WAYLAND_DEBUG`?

**Comment by spl237** [comment]
I can't see any enter or leave events at all with WAYLAND_DEBUG when moving between items in a menu. I only see them when moving into or out of a menu.

**Comment by ammen99** [comment]
> I can't see any enter or leave events at all with WAYLAND_DEBUG when moving between items in a menu. I only see them when moving into or out of a menu.

Yeah, that's expected. WAYLAND_DEBUG lets you know what Wayfire sends to the app - enter the menu surface (Wayfire doesn't know about individual menu items, only about the menu surface as a whole) and motion events. If everything here works, I'm not sure what Wayfire can be doing wrong to block GTK like this, especially since the other menu items work - hence I really suspect a GTK bug. Have you tried asking the GTK devs about this?

**Comment by spl237** [comment]
> Have you tried asking the GTK devs about this?

No - I'm not sure how interested they will be in GTK3 issues any more...

**Comment by spl237** [comment]
After a lot of digging about in the GTK code, I can confirm that this does indeed look like a GTK bug. I eventually tracked down the code which generates the enter notify event (which was not easy...), and I can fix the problem by commenting out a test in the relevant function in gdkwindow.c, but that is almost certainly not the right thing to do.

The cause appears to be that windows are storing lists of devices which are currently in that window; when a device leaves or enters a window, the list of devices is checked to see if the device is newly added to or newly removed from the window in question, and the event is only generated if that is the case. Under X this seems to work, but when a menu entry is clicked, the list of devices supplied is empty (but should not be), so when the device is removed, the event is not generated.

I have absolutely no idea how to fix this!

**Comment by spl237** [comment]
OK, confirmed as GTK bug and fix found.

The devices_inside list for a GDK window stores the pointers which are currently inside that window, and as far as I can tell is only used to detect when a pointer enters or leaves the window, generating the relevant enter and leave notification events.

When a menu is closed in X, the devices_inside list for the menu item which was clicked has the current device removed as part of the process. Under Wayland, the list is not cleared when the item is clicked, so when the menu is reopened, the pointer is viewed as already being inside that item, so no enter notification event is generated, so no prelight.

After a bit of thought, I think it is much easier to simply free the devices_inside list and null its pointer whenever gdk_window_hide is called - I can't see any circumstances in which it is valid to keep a window's devices_inside list when it is hidden. I've locally patched GTK to do that, and it seems to have fixed the problem.

There is probably a more correct fix in terms of actually working out why the mechanism which should be removing the pointer from the list isn't doing so under Wayland, but life is far too short for that - I've already spent two days on this!

Closing as definitely not a Wayfire bug. Hopefully the info above will help anyone else who encounters it...
